File: algorithm/java/Java_Algorithm/src/Safety2468/Safety2468bfs.java
package Safety2468;

import java.io.*;
import java.util.*;

public class Safety2468bfs {

    static int N, maxCnt, minH, maxH;
    static int[][] map;
    static boolean[][] visited;
    static int dr[] = {0,1,0,-1};
    static int dc[] = {1,0,-1,0};

    static class Pair {
        public int row, col;

        public Pair(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        
        map = new int[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer stk = new StringTokenizer(br.readLine(), " ");
            int index = 0;

            while(stk.hasMoreTokens()) {
                int height = Integer.parseInt(stk.nextToken());
                maxH = Math.max(height, maxH);
                minH = Math.min(height, minH);

                map[i][index++] = height;
            }
        }
        
        for (int h = minH-1; h < maxH; h++) {
            
            int cnt = 0;
            
            for (int i = 0; i < N*N; i++) {
                int row = i / N;
                int col = i % N;

                if (map[row][col] > h && !visited[row][col]) {
                    bfs(row, col, h);
                    cnt++;
                }
            }
            
            maxCnt = Math.max(maxCnt, cnt);
            visited = new boolean[N][N];
        }


        bw.write(maxCnt + "");
        bw.close();
        br.close(); 
    }

    static void bfs(int row, int col, int height) {
        visited[row][col] = true;
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(row, col));

        while(!queue.isEmpty()) {

            int curR = queue.peek().row;
            int curC = queue.peek().col;
            queue.poll();

            for (int i = 0; i < 4; i++) {
                int newR = curR + dr[i];
                int newC = curC + dc[i];
                
                if (newR >= 0 && newC >= 0 && newR < N && newC < N && !visited[newR][newC] && map[newR][newC] > height) {
                    queue.add(new Pair(newR, newC));
                    visited[newR][newC] = true;
                }
            }
        }
    }
}

File: algorithm/python/Safety2468/safety2468_bfs.py
import sys
from queue import Queue

input = sys.stdin.readline

dr = [0,1,0,-1]
dc = [1,0,-1,0]

N = int(input())

max_cnt = 0
min_h = 0
max_h = 0
visited = [[0] * N for _ in range(N)]
safety_map = []


def bfs(row, col, h):
    
    visited[row][col] = 1
    queue = Queue()
    queue.put((row, col))

    while not queue.empty():

        cur_row, cur_col = queue.get()
        
        for i in range(0,4):
            new_r = cur_row + dr[i]
            new_c = cur_col + dc[i]

            if (new_r >= 0 and new_c >= 0 and new_r < N and new_c < N and visited[new_r][new_c] == 0 and safety_map[new_r][new_c] > h):
                queue.put((new_r,new_c))
                visited[new_r][new_c] = 1    


for i in range(N):
    numbers = list(map(int, input().split()))
    
    for j in range(len(numbers)):
        min_h = min(numbers[j], min_h)
        max_h = max(numbers[j], max_h)
    
    safety_map.append(numbers)


for h in range(min_h-1, max_h):
    
    cur_cnt = 0

    for i in range(N*N):
        row = i // N
        col = i % N
        
        if (safety_map[row][col] > h and visited[row][col] == 0):
            bfs(row,col,h)
            cur_cnt += 1

    visited = [[0] * N for _ in range(N)]
    max_cnt = max(max_cnt, cur_cnt)

print(max_cnt)

